gpflow/optimizers/scipy.py

```diff
@@ -99,34 +100,37 @@ def _eval(x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
 
     @classmethod
     def callback_func(
-        cls, variables: Variables, step_callback: StepCallback
+        cls, variables: List[tf.Variable], step_callback: StepCallback
     ) -> Callable[[np.ndarray], None]:
         step = 0  # type: int
 
         def _callback(x: np.ndarray) -> None:
             nonlocal step
             values = cls.unpack_tensors(variables, x)
-            step_callback(step=step, variables=variables, values=values)
+            step_callback(step, variables, values)
             step += 1
 
         return _callback
 
     @staticmethod
-    def pack_tensors(tensors: Iterable[tf.Tensor]) -> tf.Tensor:
+    def pack_tensors(tensors: Iterable[Union[tf.Tensor, tf.Variable]]) -> tf.Tensor:
         flats = [tf.reshape(tensor, (-1,)) for tensor in tensors]
         tensors_vector = tf.concat(flats, axis=0)
         return tensors_vector
 
     @staticmethod
-    def unpack_tensors(to_tensors: Iterable[tf.Tensor], from_vector: tf.Tensor) -> List[tf.Tensor]:
+    def unpack_tensors(
+        to_tensors: Iterable[Union[tf.Tensor, tf.Variable]], from_vector: tf.Tensor
+    ) -> List[tf.Tensor]:
         s = 0
         values = []
-        for tensor in to_tensors:
-            shape = tf.shape(tensor)
+        for target_tensor in to_tensors:
+            shape = tf.shape(target_tensor)
+            dtype = target_tensor.dtype
             tensor_size = tf.reduce_prod(shape)
-            tensor_vector = tf.cast(from_vector[s : s + tensor_size], tensor.dtype)
-            tensor_vector = tf.reshape(tensor_vector, shape)
-            values.append(tensor_vector)
+            tensor_vector = from_vector[s : s + tensor_size]
+            tensor = tf.reshape(tf.cast(tensor_vector, dtype), shape)
+            values.append(tensor)
```

Comment:
```
        s = 0
        v = []
        for t in to_tensors:
            s_ = tf.shape(t)
            d = t.dtype
            s__ = tf.reduce_prod(s_)
            t_ = from_vector[s : s + s__]
            t__ = tf.reshape(tf.cast(t_, d) s_)
            v.append(t__)
            s += s__
        return v
```

Reply:
I found the original code hard to parse, particularly because of the re-assignment to the same variable. I've tried to clarify what is happening here by separating the splicing of the right subvector from turning that vector into the fully shaped&typed tensor again. If you have suggestions for further improvements I'd be happy to incorporate them!